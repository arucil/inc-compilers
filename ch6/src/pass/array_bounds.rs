use ast::{Error, Exp, ExpKind, FuncDef, IdxVar, Program, Type};
use id_arena::Arena;
use support::Range;

/// (make-vector e1 e2)
/// =>
/// (let ([len e1])
///   (if (>= len 0)
///     (make-vector len e2)
///     (length-error)))
///
/// (vector-ref e1 e2)
/// =>
/// (let ([vec e1] [idx e2])
///   (if (and (>= idx 0)
///            (< idx (vector-length vec)))
///     (vector-ref vec idx)
///     (out-of-bounds idx (vector-length vec))))
///
/// (vector-set! e1 e2 e3)
/// =>
/// (let ([vec e1] [idx e2])
///   (if (and (>= idx 0) (< idx (vector-length vec)))
///     (vector-set! vec idx e3)
///     (out-of-bounds idx (vector-length vec))))
pub fn insert_bounds_check(
  prog: Program<IdxVar, Type>,
) -> Program<IdxVar, Type> {
  let mut state = State {
    types: &prog.types,
    tmp_counter: 0,
  };
  Program {
    func_defs: prog
      .func_defs
      .into_iter()
      .map(|(name, func)| {
        (
          name,
          FuncDef {
            body: state.exp_insert(func.body),
            ..func
          },
        )
      })
      .collect(),
    body: prog
      .body
      .into_iter()
      .map(|exp| state.exp_insert(exp))
      .collect(),
    ..prog
  }
}

struct State<'a> {
  types: &'a Arena<Type>,
  tmp_counter: usize,
}

impl<'a> State<'a> {
  fn exp_insert(&mut self, exp: Exp<IdxVar, Type>) -> Exp<IdxVar, Type> {
    match exp.kind {
      ExpKind::Int(_)
      | ExpKind::Var(_)
      | ExpKind::Str(_)
      | ExpKind::Bool(_)
      | ExpKind::Get(_)
      | ExpKind::Void
      | ExpKind::NewLine => exp,
      ExpKind::Prim {
        op: op @ (_, "make-vector"),
        mut args,
      } => {
        let range = exp.range;
        let ty = exp.ty;
        let val = self.exp_insert(args.pop().unwrap());
        let len = self.exp_insert(args.pop().unwrap());
        self.make_check_exp(
          range,
          ty.clone(),
          vec![],
          len,
          |len_var| Exp {
            kind: ExpKind::Prim {
              op: (range, ">="),
              args: vec![
                len_var,
                Exp {
                  kind: ExpKind::Int(0),
                  range,
                  ty: Type::Int,
                },
              ],
            },
            range,
            ty: Type::Bool,
          },
          |len_var| Exp {
            kind: ExpKind::Prim {
              op,
              args: vec![len_var, val],
            },
            range,
            ty,
          },
          |len_var| Error::Length(box len_var),
        )
      }
      ExpKind::Prim {
        op: op @ (_, "vector-ref"),
        mut args,
      } if matches!(self.resolve_type(&args[0].ty), Type::Array(_)) => {
        let range = exp.range;
        let ty = exp.ty;
        let idx = self.exp_insert(args.pop().unwrap());
        let vec = self.exp_insert(args.pop().unwrap());
        let mut tmps = vec![];
        let vec_var;
        if vec.kind.is_atomic() {
          vec_var = vec;
        } else {
          let tmp = self.new_tmp();
          vec_var = Exp {
            kind: ExpKind::Var(tmp.clone()),
            range: vec.range,
            ty: vec.ty.clone(),
          };
          tmps.push(((vec.range, tmp), box vec));
        }
        self.make_check_exp(
          range,
          ty.clone(),
          tmps,
          idx,
          |idx_var| Exp {
            kind: ExpKind::Prim {
              op: (range, "and"),
              args: vec![
                Exp {
                  kind: ExpKind::Prim {
                    op: (range, ">="),
                    args: vec![
                      idx_var.clone(),
                      Exp {
                        kind: ExpKind::Int(0),
                        range,
                        ty: Type::Int,
                      },
                    ],
                  },
                  range,
                  ty: Type::Bool,
                },
                Exp {
                  kind: ExpKind::Prim {
                    op: (range, "<"),
                    args: vec![
                      idx_var,
                      Exp {
                        kind: ExpKind::Prim {
                          op: (range, "vector-length"),
                          args: vec![vec_var.clone()],
                        },
                        range,
                        ty: Type::Int,
                      },
                    ],
                  },
                  range,
                  ty: Type::Bool,
                },
              ],
            },
            range,
            ty: Type::Bool,
          },
          |idx_var| Exp {
            kind: ExpKind::Prim {
              op,
              args: vec![vec_var.clone(), idx_var],
            },
            range,
            ty: ty.clone(),
          },
          |idx_var| Error::OutOfBounds {
            index: box idx_var,
            len: box Exp {
              kind: ExpKind::Prim {
                op: (range, "vector-length"),
                args: vec![vec_var.clone()],
              },
              range,
              ty: Type::Int,
            },
          },
        )
      }
      ExpKind::Prim {
        op: op @ (_, "vector-set!"),
        mut args,
      } if matches!(self.resolve_type(&args[0].ty), Type::Array(_)) => {
        let range = exp.range;
        let ty = exp.ty;
        let val = self.exp_insert(args.pop().unwrap());
        let idx = self.exp_insert(args.pop().unwrap());
        let vec = self.exp_insert(args.pop().unwrap());
        let mut tmps = vec![];
        let vec_var;
        if vec.kind.is_atomic() {
          vec_var = vec;
        } else {
          let tmp = self.new_tmp();
          vec_var = Exp {
            kind: ExpKind::Var(tmp.clone()),
            range: vec.range,
            ty: vec.ty.clone(),
          };
          tmps.push(((vec.range, tmp), box vec));
        }
        self.make_check_exp(
          range,
          ty.clone(),
          tmps,
          idx,
          |idx_var| Exp {
            kind: ExpKind::Prim {
              op: (range, "and"),
              args: vec![
                Exp {
                  kind: ExpKind::Prim {
                    op: (range, ">="),
                    args: vec![
                      idx_var.clone(),
                      Exp {
                        kind: ExpKind::Int(0),
                        range,
                        ty: Type::Int,
                      },
                    ],
                  },
                  range,
                  ty: Type::Bool,
                },
                Exp {
                  kind: ExpKind::Prim {
                    op: (range, "<"),
                    args: vec![
                      idx_var,
                      Exp {
                        kind: ExpKind::Prim {
                          op: (range, "vector-length"),
                          args: vec![vec_var.clone()],
                        },
                        range,
                        ty: Type::Int,
                      },
                    ],
                  },
                  range,
                  ty: Type::Bool,
                },
              ],
            },
            range,
            ty: Type::Bool,
          },
          |idx_var| Exp {
            kind: ExpKind::Prim {
              op,
              args: vec![vec_var.clone(), idx_var, val],
            },
            range,
            ty: ty.clone(),
          },
          |idx_var| Error::OutOfBounds {
            index: box idx_var,
            len: box Exp {
              kind: ExpKind::Prim {
                op: (range, "vector-length"),
                args: vec![vec_var.clone()],
              },
              range,
              ty: Type::Int,
            },
          },
        )
      }
      ExpKind::Prim { op, args } => Exp {
        kind: ExpKind::Prim {
          op,
          args: args.into_iter().map(|exp| self.exp_insert(exp)).collect(),
        },
        ..exp
      },
      ExpKind::Apply {
        func,
        args,
        r#struct,
      } => Exp {
        kind: ExpKind::Apply {
          func: box self.exp_insert(*func),
          args: args.into_iter().map(|exp| self.exp_insert(exp)).collect(),
          r#struct,
        },
        ..exp
      },
      ExpKind::Let { var, init, body } => Exp {
        kind: ExpKind::Let {
          var,
          init: box self.exp_insert(*init),
          body: box self.exp_insert(*body),
        },
        ..exp
      },
      ExpKind::If { cond, conseq, alt } => Exp {
        kind: ExpKind::If {
          cond: box self.exp_insert(*cond),
          conseq: box self.exp_insert(*conseq),
          alt: box self.exp_insert(*alt),
        },
        ..exp
      },
      ExpKind::Set { var, exp: exp1 } => Exp {
        kind: ExpKind::Set {
          var,
          exp: box self.exp_insert(*exp1),
        },
        ..exp
      },
      ExpKind::Begin { seq, last } => Exp {
        kind: ExpKind::Begin {
          seq: seq.into_iter().map(|exp| self.exp_insert(exp)).collect(),
          last: box self.exp_insert(*last),
        },
        ..exp
      },
      ExpKind::While { cond, body } => Exp {
        kind: ExpKind::While {
          cond: box self.exp_insert(*cond),
          body: box self.exp_insert(*body),
        },
        ..exp
      },
      ExpKind::Print(args) => Exp {
        kind: ExpKind::Print(
          args.into_iter().map(|exp| self.exp_insert(exp)).collect(),
        ),
        ..exp
      },
      ExpKind::Error(_) => unreachable!(),
      ExpKind::FunRef { .. } => exp,
    }
  }

  fn make_check_exp<F, G, E>(
    &mut self,
    range: Range,
    ty: Type,
    mut tmps: Vec<Tmp>,
    index: Exp<IdxVar, Type>,
    cond: F,
    conseq: G,
    err: E,
  ) -> Exp<IdxVar, Type>
  where
    F: FnOnce(Exp<IdxVar, Type>) -> Exp<IdxVar, Type>,
    G: FnOnce(Exp<IdxVar, Type>) -> Exp<IdxVar, Type>,
    E: FnOnce(Exp<IdxVar, Type>) -> Error<IdxVar, Type>,
  {
    let idx_var;
    if index.kind.is_atomic() {
      idx_var = index;
    } else {
      let tmp = self.new_tmp();
      idx_var = Exp {
        kind: ExpKind::Var(tmp.clone()),
        range: index.range,
        ty: Type::Int,
      };
      tmps.push(((index.range, tmp), box index));
    }
    let body = Exp {
      kind: ExpKind::If {
        cond: box cond(idx_var.clone()),
        conseq: box conseq(idx_var.clone()),
        alt: box Exp {
          kind: ExpKind::Error(err(idx_var)),
          range,
          ty: ty.clone(),
        },
      },
      range,
      ty: ty.clone(),
    };
    tmps.into_iter().rfold(body, |body, (var, init)| Exp {
      kind: ExpKind::Let {
        var,
        init,
        body: box body,
      },
      range,
      ty: ty.clone(),
    })
  }

  fn resolve_type(&self, ty: &Type) -> Type {
    ty.resolved(self.types)
  }

  fn new_tmp(&mut self) -> IdxVar {
    self.tmp_counter += 1;
    IdxVar {
      name: "(bounds-tmp)".to_owned(),
      index: self.tmp_counter,
    }
  }
}

type Tmp = ((Range, IdxVar), Box<Exp<IdxVar, Type>>);

#[cfg(test)]
mod tests {
  use super::*;
  use ast::*;
  use ch4::pass::uniquify;
  use ch5::pass::typecheck::typecheck;
  use insta::assert_snapshot;

  #[test]
  fn array() {
    let prog = parse(
      r#"
(let ([x (make-vector 3 #t)]
      [i 1]
      [y (vector-ref x i)]
      [z (vector 1 #t "abc")])
  (vector-set! x i (vector-ref z 1))
  (vector-set! z 1 (not y)))
      "#,
    )
    .unwrap();
    let prog = typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let result = insert_bounds_check(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
