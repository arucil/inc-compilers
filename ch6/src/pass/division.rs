use ast::{Error, Exp, ExpKind, Program, Type};
use id_arena::Arena;
use support::Range;

pub fn insert_division_check(
  prog: Program<String, Type>,
) -> Program<String, Type> {
  let types = &prog.types;
  Program {
    body: prog
      .body
      .into_iter()
      .map(|exp| exp_insert(exp, types))
      .collect(),
    ..prog
  }
}