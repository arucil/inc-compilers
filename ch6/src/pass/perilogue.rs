use super::register_allocation::Info;
use asm::{Arg, Block, Instr, Label, Program};

pub fn add_perilogue(mut prog: Program<Info>) -> Program<Info> {
  add_prologue(&mut prog);
  add_epilogue(&mut prog);
  prog
}

fn add_prologue(prog: &mut Program<Info>) {
  use asm::Reg::*;
  use Instr::*;
  let stack_space =
    prog.info.stack_space + prog.info.used_callee_saved_regs.len() * 8;
  let stack_space =
    ((stack_space + 15) & !15) - prog.info.used_callee_saved_regs.len() * 8;
  let mut code = vec![
    Push(Arg::Reg(Rbp)),
    Mov {
      src: Arg::Reg(Rsp),
      dest: Arg::Reg(Rbp),
    },
    Sub {
      src: Arg::Imm(stack_space as i64),
      dest: Arg::Reg(Rsp),
    },
  ];
  for &reg in &prog.info.used_callee_saved_regs {
    code.push(Push(Arg::Reg(reg)));
  }
  code.extend_from_slice(&[
    Mov {
      src: Arg::Imm(65536),
      dest: Arg::Reg(Rdi),
    },
    Mov {
      src: Arg::Imm(64),
      dest: Arg::Reg(Rsi),
    },
    Call {
      label: "rt_initialize".to_owned(),
      arity: 2,
    },
    Mov {
      src: Arg::Label("rt_rootstack_begin".to_owned()),
      dest: Arg::Reg(R15),
    },
  ]);
  for i in (0..prog.info.rootstack_space as i32).step_by(8) {
    code.push(Mov {
      src: Arg::Imm(0),
      dest: Arg::Deref(R15, i),
    });
  }
  code.extend_from_slice(&[
    Add {
      src: Arg::Imm(prog.info.rootstack_space as i64),
      dest: Arg::Reg(R15),
    },
    Jmp(Label::Start),
  ]);
  let block = Block { global: true, code };
  prog.blocks.push((Label::EntryPoint, block));
}

fn add_epilogue(prog: &mut Program<Info>) {
  use asm::Reg::*;
  use Instr::*;
  let mut code: Vec<Instr> = vec![Sub {
    src: Arg::Imm(prog.info.rootstack_space as i64),
    dest: Arg::Reg(R15),
  }];
  code.extend(
    prog
      .info
      .used_callee_saved_regs
      .iter()
      .rev()
      .map(|&reg| Pop(Arg::Reg(reg))),
  );
  code.extend_from_slice(&[
    Mov {
      src: Arg::Reg(Rbp),
      dest: Arg::Reg(Rsp),
    },
    Pop(Arg::Reg(Rbp)),
    Mov {
      src: Arg::Imm(60),
      dest: Arg::Reg(Rax),
    },
    Mov {
      src: Arg::Imm(0),
      dest: Arg::Reg(Rdi),
    },
    Syscall,
  ]);
  let block = Block {
    global: false,
    code,
  };
  prog.blocks.push((Label::Conclusion, block));
}
