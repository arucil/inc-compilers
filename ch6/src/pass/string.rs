use ast::{Error, Exp, ExpKind, FunDef, IdxVar, Program, Type};

pub fn expose_string_concat(
  prog: Program<IdxVar, Type>,
) -> Program<IdxVar, Type> {
  let mut state = State { tmp_counter: 0 };
  Program {
    fun_defs: prog
      .fun_defs
      .into_iter()
      .map(|(name, fun)| {
        (
          name,
          FunDef {
            body: state.exp_insert(fun.body),
            ..fun
          },
        )
      })
      .collect(),
    body: state.exp_insert(prog.body),
    ..prog
  }
}

struct State {
  tmp_counter: usize,
}

impl State {
  /// (string-append e1 e2)
  /// =>
  /// (let ([tmp1 e1]
  ///       [tmp2 e2]
  ///       [len1 (string-length tmp1)]
  ///       [result (alloc-string
  ///                 (+ len1 (string-length tmp2)))])
  ///   (copy-string! result 0 tmp1)
  ///   (copy-string! result len1 tmp2))
  fn exp_insert(&mut self, exp: Exp<IdxVar, Type>) -> Exp<IdxVar, Type> {
    match exp.kind {
      ExpKind::Int(_)
      | ExpKind::Var(_)
      | ExpKind::Str(_)
      | ExpKind::Bool(_)
      | ExpKind::Get(_)
      | ExpKind::Void
      | ExpKind::NewLine => exp,
      ExpKind::Prim {
        op: (_, "string-append"),
        mut args,
      } => {
        let range = exp.range;
        let arg2 = self.exp_insert(args.pop().unwrap());
        let arg1 = self.exp_insert(args.pop().unwrap());
        let tmp1 = self.new_tmp();
        let arg1_var = Exp {
          kind: ExpKind::Var(tmp1.clone()),
          range: exp.range,
          ty: Type::Int,
        };
        let tmp2 = self.new_tmp();
        let arg2_var = Exp {
          kind: ExpKind::Var(tmp2.clone()),
          range: exp.range,
          ty: Type::Int,
        };
        let len1 = self.new_tmp();
        let len1_var = Exp {
          kind: ExpKind::Var(len1.clone()),
          range: exp.range,
          ty: Type::Int,
        };
        let tmp3 = self.new_tmp();
        let result_var = Exp {
          kind: ExpKind::Var(tmp3.clone()),
          range: exp.range,
          ty: Type::Int,
        };
        let body = Exp {
          kind: ExpKind::Begin {
            seq: vec![
              Exp {
                kind: ExpKind::Prim {
                  op: (range, "copy-string!"),
                  args: vec![
                    result_var.clone(),
                    Exp {
                      kind: ExpKind::Int(0),
                      range,
                      ty: Type::Int,
                    },
                    arg1_var.clone(),
                  ],
                },
                range,
                ty: Type::Void,
              },
              Exp {
                kind: ExpKind::Prim {
                  op: (range, "copy-string!"),
                  args: vec![
                    result_var.clone(),
                    len1_var.clone(),
                    arg2_var.clone(),
                  ],
                },
                range,
                ty: Type::Void,
              },
            ],
            last: box result_var,
          },
          range,
          ty: Type::Str,
        };
        vec![
          (tmp1, box arg1),
          (tmp2, box arg2),
          (
            len1,
            box Exp {
              kind: ExpKind::Prim {
                op: (range, "string-length"),
                args: vec![arg1_var],
              },
              range,
              ty: Type::Int,
            },
          ),
          (
            tmp3,
            box Exp {
              kind: ExpKind::Prim {
                op: (exp.range, "alloc-string"),
                args: vec![Exp {
                  kind: ExpKind::Prim {
                    op: (exp.range, "+"),
                    args: vec![
                      len1_var,
                      Exp {
                        kind: ExpKind::Prim {
                          op: (range, "string-length"),
                          args: vec![arg2_var],
                        },
                        range,
                        ty: Type::Int,
                      },
                    ],
                  },
                  range: exp.range,
                  ty: Type::Int,
                }],
              },
              range,
              ty: Type::Str,
            },
          ),
        ]
        .into_iter()
        .rfold(body, |body, (var, init)| Exp {
          kind: ExpKind::Let {
            var: (range, var),
            init,
            body: box body,
          },
          range,
          ty: Type::Str,
        })
      }
      ExpKind::Prim { op, args } => Exp {
        kind: ExpKind::Prim {
          op,
          args: args.into_iter().map(|exp| self.exp_insert(exp)).collect(),
        },
        ..exp
      },
      ExpKind::Apply {
        fun,
        args,
        r#struct,
      } => Exp {
        kind: ExpKind::Apply {
          fun: box self.exp_insert(*fun),
          args: args.into_iter().map(|exp| self.exp_insert(exp)).collect(),
          r#struct,
        },
        ..exp
      },
      ExpKind::Let { var, init, body } => Exp {
        kind: ExpKind::Let {
          var,
          init: box self.exp_insert(*init),
          body: box self.exp_insert(*body),
        },
        ..exp
      },
      ExpKind::If { cond, conseq, alt } => Exp {
        kind: ExpKind::If {
          cond: box self.exp_insert(*cond),
          conseq: box self.exp_insert(*conseq),
          alt: box self.exp_insert(*alt),
        },
        ..exp
      },
      ExpKind::Set { var, exp: exp1 } => Exp {
        kind: ExpKind::Set {
          var,
          exp: box self.exp_insert(*exp1),
        },
        ..exp
      },
      ExpKind::Begin { seq, last } => Exp {
        kind: ExpKind::Begin {
          seq: seq.into_iter().map(|exp| self.exp_insert(exp)).collect(),
          last: box self.exp_insert(*last),
        },
        ..exp
      },
      ExpKind::While { cond, body } => Exp {
        kind: ExpKind::While {
          cond: box self.exp_insert(*cond),
          body: box self.exp_insert(*body),
        },
        ..exp
      },
      ExpKind::Print(args) => Exp {
        kind: ExpKind::Print(
          args.into_iter().map(|exp| self.exp_insert(exp)).collect(),
        ),
        ..exp
      },
      ExpKind::Error(Error::Length(arg)) => Exp {
        kind: ExpKind::Error(Error::Length(box self.exp_insert(*arg))),
        ..exp
      },
      ExpKind::Error(Error::OutOfBounds { index, len }) => Exp {
        kind: ExpKind::Error(Error::OutOfBounds {
          index: box self.exp_insert(*index),
          len: box self.exp_insert(*len),
        }),
        ..exp
      },
      ExpKind::Error(Error::DivByZero) => exp,
      ExpKind::FunRef { .. } => exp,
    }
  }

  fn new_tmp(&mut self) -> IdxVar {
    self.tmp_counter += 1;
    IdxVar {
      name: "(str-tmp)".to_owned(),
      index: self.tmp_counter,
    }
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use ast::*;
  use ch4::pass::uniquify;
  use ch5::pass::typecheck::typecheck;
  use insta::assert_snapshot;

  #[test]
  fn string_append() {
    let prog = parse(
      r#"
(let ([x "\n"])
  (print (string-append "abc" (string-append "+" x))))
      "#,
    )
    .unwrap();
    let prog = typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let result = expose_string_concat(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
