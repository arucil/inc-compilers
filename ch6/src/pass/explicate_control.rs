#[cfg(test)]
mod tests {
  use ch2::pass::remove_complex_operands;
  use ch4::pass::explicate_control;
  use ch4::pass::shrink;
  use ch4::pass::uniquify;
  use ch5::pass::typecheck;
  use insta::assert_snapshot;

  #[test]
  fn vector() {
    let prog = ast::parse(
      r#"
(let ([x (void)] [y (vector 1 #f (void) (vector (set! x x)) "abc")])
  (set! x (vector-set! y 1 (not (vector-ref y 1))))
  (set! x (vector-ref y 2))
  (vector-set! y 0 (vector-length y))
  (vector-set! y 2 (vector-ref (vector-ref y 3) 0)))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control::explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn r#struct() {
    let prog = ast::parse(
      r#"
(define-struct my-void (x Void))
(define-struct my-comp (x Int) (y Bool) (z Void) (u my-void) (v Str))
(let ([x (void)] [y (my-comp 1 #f (void) (my-void (set! x x)) "abc")])
  (set! x (set-my-comp-y! y (not (my-comp-y y))))
  (set! x (my-comp-z y))
  (set-my-comp-x! y (vector-length y))
  (set-my-comp-z! y (my-void-x (my-comp-u y))))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control::explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
