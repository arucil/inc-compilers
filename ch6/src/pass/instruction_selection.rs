#[cfg(test)]
mod tests {
  use ch2::pass::remove_complex_operands;
  use ch4::pass::explicate_control;
  use ch4::pass::instruction_selection;
  use ch4::pass::uniquify;
  use ch5::pass::typecheck;
  use insta::assert_snapshot;

  #[test]
  fn vector() {
    let prog = ast::parse(
      r#"
(let ([x (void)] [y (vector 1 #f (void) (vector (set! x x)) "abc")])
  (set! x (vector-set! y 1 (not (vector-ref y 1))))
  (set! x (vector-ref y 2))
  (vector-set! y 0 (vector-length y))
  (vector-set! y 2 (vector-ref (vector-ref y 3) 0)))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let prog = explicate_control::explicate_control(prog);
    let result = instruction_selection::select_instruction(prog, true);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn empty_struct() {
    let prog = ast::parse(
      r#"
(define-struct foo)
(let ([x (foo)])
  (void))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let prog = explicate_control::explicate_control(prog);
    let result = instruction_selection::select_instruction(prog, true);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn array() {
    let prog = ast::parse(
      r#"
(let ([x (make-vector 3 #t)]
      [y (make-vector 0 2)]
      [z (make-vector 2 x)]
      [t (make-vector 10 (void))]
      [u (void)])
  (vector-set! x 1 #f)
  (vector-set! t 3 (void))
  (set! u (vector-ref t (vector-ref y 0))))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let prog = explicate_control::explicate_control(prog);
    let result = instruction_selection::select_instruction(prog, true);
    assert_snapshot!(result.to_string_pretty());
  }
}
