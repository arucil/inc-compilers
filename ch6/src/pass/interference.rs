#[cfg(test)]
mod tests {
  use asm::{Label, Program};
  use ast::{IdxVar, Type};
  use ch3::location_set::LocationSet;
  use ch4::pass::instruction_selection::Info as OldOldInfo;
  use ch4::pass::interference;
  use ch4::pass::liveness_analysis;
  use indexmap::indexmap;
  use insta::assert_snapshot;
  use maplit::hashmap;

  #[test]
  fn vector() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      pop rdi
      pop rsi
      push x
      mov w, rbx
      call rt_allocate, 2, gc
      add w, x
      jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexmap! {
          IdxVar::new("x") => Type::Tuple(vec![Type::Int]),
          IdxVar::new("w") => Type::Int,
        },
      },
      blocks,
      ..Program::default()
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = interference::build_interference(prog);

    assert_snapshot!(format!("{}", result.info));
  }

  #[test]
  fn str() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      pop rdi
      pop rsi
      push x
      mov w, rbx
      call rt_allocate, 2, gc
      add w, x
      jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexmap! {
          IdxVar::new("x") => Type::Str,
          IdxVar::new("w") => Type::Int,
        },
      },
      blocks,
      ..Program::default()
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let result = interference::build_interference(prog);

    assert_snapshot!(format!("{}", result.info));
  }
}
