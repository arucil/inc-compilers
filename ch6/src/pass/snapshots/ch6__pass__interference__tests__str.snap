---
source: ch6/src/pass/interference.rs
expression: "format!(\"{:?}\", result.info)"
---
graph {
    0 [ label = "x.0" ]
    1 [ label = "w.0" ]
    2 [ label = "rdi" ]
    3 [ label = "rsp" ]
    4 [ label = "rbx" ]
    5 [ label = "rsi" ]
    6 [ label = "rax" ]
    7 [ label = "rcx" ]
    8 [ label = "rdx" ]
    9 [ label = "r8" ]
    10 [ label = "r9" ]
    11 [ label = "r10" ]
    12 [ label = "r11" ]
    13 [ label = "rbp" ]
    14 [ label = "r12" ]
    15 [ label = "r13" ]
    16 [ label = "r14" ]
    17 [ label = "r15" ]
    2 -- 3 [ ]
    2 -- 4 [ ]
    2 -- 0 [ ]
    5 -- 3 [ ]
    5 -- 4 [ ]
    5 -- 2 [ ]
    5 -- 0 [ ]
    1 -- 3 [ ]
    1 -- 5 [ ]
    1 -- 2 [ ]
    1 -- 0 [ ]
    6 -- 3 [ ]
    6 -- 0 [ ]
    6 -- 1 [ ]
    7 -- 3 [ ]
    7 -- 6 [ ]
    7 -- 0 [ ]
    7 -- 1 [ ]
    8 -- 3 [ ]
    8 -- 6 [ ]
    8 -- 0 [ ]
    8 -- 1 [ ]
    5 -- 6 [ ]
    2 -- 6 [ ]
    9 -- 3 [ ]
    9 -- 6 [ ]
    9 -- 0 [ ]
    9 -- 1 [ ]
    10 -- 3 [ ]
    10 -- 6 [ ]
    10 -- 0 [ ]
    10 -- 1 [ ]
    11 -- 3 [ ]
    11 -- 6 [ ]
    11 -- 0 [ ]
    11 -- 1 [ ]
    12 -- 3 [ ]
    12 -- 6 [ ]
    12 -- 0 [ ]
    12 -- 1 [ ]
    3 -- 0 [ ]
    13 -- 0 [ ]
    4 -- 0 [ ]
    14 -- 0 [ ]
    15 -- 0 [ ]
    16 -- 0 [ ]
    17 -- 0 [ ]
}

